zap/src/irgen/ser.rs

```diff
@@ -186,19 +266,38 @@ impl Ser<'_> {
 
 		if optional {
 			self.push_stmt(Stmt::ElseIf(Expr::from(from.clone()).eq(Expr::Nil)));
-			self.push_writenumty((i_offset as f64).into(), discriminant_numty);
+			self.push_variant_storage(&storage, i_offset);
 		}
 
 		self.push_stmt(Stmt::Else);
 		if let Some(unknown_i) = unknown_i {
-			self.push_writenumty(Expr::from(unknown_i as f64), discriminant_numty);
+			self.push_variant_storage(&storage, unknown_i);
 			self.push_ty(&Ty::Unknown, from.clone());
 		} else {
 			self.push_stmt(Stmt::Error("Invalid type".into()));
 		}
 		self.push_stmt(Stmt::End);
 	}
 
+	fn set_bitfield(&mut self, var: Var, bits: BitpackMask) {
+		self.push_assign(
+			var.clone(),
+			Expr::Call(
+				Var::NameIndex(Var::Name("bit32".into()).into(), "bor".into()).into(),
+				None,
+				vec![Expr::Var(var.into()), Expr::BinaryNum(bits)],
+			),
+		);
+	}
+
+	fn push_bool<F: FnOnce(&mut Self)>(&mut self, cond_expr: Expr, cb: F) {
+		let (bits, var) = self.get_bitpack();
+		self.push_stmt(Stmt::If(cond_expr));
+		self.set_bitfield(var, bits);
+		cb(self);
+		self.push_stmt(Stmt::End);
+	}
```

Comment:
Don't require a conditional expression/callback here... just have `push_bool` have `self.get_bitpack` and `self.set_bitfield`.